### ISSUE TITLE
no-unmodified-loop-condition too restrictive

### ISSUE THREAD

**Issue by danielrentz** [eslint/eslint#6280]
**What version of ESLint are you using?**
2.11.0 (via grunt-eslint 18.1.0)

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

**What did you do? Please include the actual source code causing the issue.**
Minified example:

    function checkThis() { ... }
    function checkThat() { ... }

    var cond = something();
    while (checkThis() || (cond && checkThat())) { ... }

**What did you expect to happen?**
Code passes ESLint.

**What actually happened? Please include the actual, raw output from ESLint.**
  error  'cond' is not modified in this loop  no-unmodified-loop-condition

**Comment by platinumazure** [comment]
Hmm. Seems we should be deciding that `checkThis`, being a function call, is potentially modifiable even if `cond` isn't.

@mysticatea What do you think? Is this a bug?

**Comment by danielrentz** [comment]
Maybe this is a better example, involving a modifiable loop variable used in the condition:

var obj = new Thing();
var cond = something();
while (obj && (obj.checkThis() || (cond && obj.checkThat()))) {
    ...
    obj = obj.next();
}

**Comment by nzakas** [comment]
I don't think this is a bug. `cond` is always the same value inside of the loop condition so it is an unmodified loop condition, and the rule is flagging it correctly. All we are doing is saying "is `cond` assigned to inside the loop?" The answer to that question is no.

**Comment by platinumazure** [comment]
@nzakas But clearly the overall loop condition is modified... Are you
suggesting that a secondary purpose of the rule might be to help flag
variables that don't need to be in the loop condition if they're constant?

I think the question comes down to "what is a loop condition"? Either it's
the entire loop test expression, or it's each condition within a complex
test. If you think the entire loop test represents the "loop condition",
then this is a bug. If it's any nested simple condition, then this is
working as designed.
On May 30, 2016 12:03 PM, "Nicholas C. Zakas" <notifications@github.com>
wrote:

I don't think this is a bug. cond is always the same value inside of the
loop condition so it is an unmodified loop condition, and the rule is
flagging it correctly. All we are doing is saying "is cond assigned to
inside the loop?" The answer to that question is no.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/eslint/eslint/issues/6280#issuecomment-222528534>, or mute
the thread
<https://github.com/notifications/unsubscribe/AARWen7osJq6r5jUhco5tnbbX12K88WGks5qGxhugaJpZM4IpkRF>
.

**Comment by mysticatea** [comment]
Hmm, this was intentional. I made this rule checking LHS and RHS of logical expressions separately. There was a case in my head:

```js
function foo(a, b) {
    while (a != null && b != null) {    /* 'b' is not modified in this loop. */
        doSomething(a, b);

        a = a.next;
        // b = b.next;
    }
}
```

But this issue is an interesting case.
Maybe we can add an option to address logical expressions in total.
`{separateLogicalExpressions: true | false}`?
And the rule name would mean that the option is `false` by default.

**Comment by mysticatea** [comment]
Or, just we can make this rule ignoring variables which are unused in the loop?

**Comment by feross** [comment]
Here's another example:

```js
var node = { index: 0 }
var sibling = 0

while (node && node.index === sibling) {
  sibling = 1
}
```

In this example, `node &&` is just an existence check.

As a user, I only care about detecting loops where the entire conditional is not modified by the body, since it is only then that I have an infinite loop.

**Comment by nzakas** [comment]
I think the larger question is: is there any reasonable way to detect these patterns or not? (And by reasonable, I mean is this going to take someone days of time to be able to catch all of these cases or is there something simple that we're just not doing?)

**Comment by platinumazure** [comment]
@feross For your particular example, you could just enclose the `while` loop in an `if` statement:

```js
var node = { index: 0 }
var sibling = 0

if (node) {
  while (node.index === sibling) {
    sibling += 1
  }
}
```

**Comment by feross** [comment]
@platinumazure Sure -- I understand users can always write their code differently. The point is that maybe this rule should only warn when *none* of the references in the loop are modified in the body.

Another way of wording this proposal is: This rule should warn when it's *sure* there's a mistake that will cause an infinite loop. The rule already misses potential infinite loops due to what's feasible to detect, i.e. getters, function calls are just assumed to potentially change the variables.

So, it seems consistent for this rule to not warn as long as one of the variables is modified.

Or, perhaps this could be an option?

**Comment by nzakas** [comment]
@feross we need someone to investigate if this is even possible to detect without a ton of added complexity, so it's not a matter of whether it's an option or not. We would be grateful for your help in that regard.

**Comment by feross** [comment]
@nzakas Seems straightforward to detect:

1. Make list of all identifiers in conditional
2. If at least one identifier is found in the body, then no warning.
3. Else, show warning.

**Comment by nzakas** [comment]
@feross high level descriptions tend to hide complexity. Can you create a pull request?

**Comment by feross** [comment]
@nzakas Would love to, but I don't have time for at least another month.

**Comment by nzakas** [comment]
Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
